mmengine/runner/deepspeed_runner.py

```diff
@@ -0,0 +1,771 @@
+                    and len(previous_gpu_ids) != self._world_size):
+                if self.model.zero_optimization():
+                    raise RuntimeError(
+                        'Cannot resuming training with ZeRO. Automatic '
+                        'adjustment of ZeRO\'s optimizer state partitioning '
+                        'with a new world size is not currently supported.'
+                        'Make sure the number of GPU is consistent with the '
+                        'previous training state resuming from the checkpoint.'
+                    )
+                else:
+                    self.logger.info(
+                        'Number of GPU used for current experiment is not '
+                        'consistent with resuming from checkpoint')
+                    if self.auto_scale_lr is None or \
+                            not self.auto_scale_lr.get('enable', False):
+                        raise RuntimeError(
+                            'Cannot automatically rescale lr in resuming. '
+                            'Please make sure the number of GPU is consistent '
+                            'with the previous training state resuming from '
+                            'the checkpoint or set `enable` in `auto_scale_lr`'
+                            ' to False.')
+
+        # resume random seed
+        resumed_seed = checkpoint['meta'].get('seed', None)
+        current_seed = self._randomness_cfg.get('seed')
+        if resumed_seed is not None and resumed_seed != current_seed:
+            if current_seed is not None:
+                warnings.warn(f'The value of random seed in the '
+                              f'checkpoint "{resumed_seed}" is '
+                              f'different from the value in '
+                              f'`randomness` config "{current_seed}"')
+            self._randomness_cfg.update(seed=resumed_seed)
+            self.set_randomness(**self._randomness_cfg)
+
+        resumed_dataset_meta = checkpoint['meta'].get('dataset_meta', None)
+        dataset_meta = getattr(self.train_dataloader.dataset, 'metainfo', None)
+        if resumed_dataset_meta != dataset_meta:
+            warnings.warn(
+                'The dataset metainfo from the resumed checkpoint is '
+                'different from the current training dataset, please '
+                'check the correctness of the checkpoint or the training '
+                'dataset.')
+
+        self.message_hub.load_state_dict(checkpoint['message_hub'])
+
+        # resume optimizer
+        if 'optimizer' in checkpoint and resume_optimizer:
+            self.optim_wrapper = self.build_optim_wrapper(self.optim_wrapper)
+            if self.model.zero_optimization():
+                self.optim_wrapper.load_state_dict(  # type: ignore
+                    checkpoint['optimizer'],
+                    load_from_fp32_weights=self.model.
+                    zero_load_from_fp32_weights())
```

Comment:
.

Reply:
This tries to follow the compatibility with https://github.com/microsoft/DeepSpeed/blob/211055216792cbb52ab6d355f698c194f9c55efb/deepspeed/runtime/zero/stage_1_and_2.py#L2228

https://github.com/microsoft/DeepSpeed/blob/211055216792cbb52ab6d355f698c194f9c55efb/deepspeed/runtime/zero/stage3.py#L2339